cachito/web/content_manifest.py

```diff
@@ -175,6 +180,44 @@ def process_rubygems_package(self, package, dependency):
             self._rubygems_data[package]["sources"].append(icm_dependency)
             self._rubygems_data[package]["dependencies"].append(icm_dependency)
 
+    def _replace_gomod_parent_dir_dependency(
+        self, package: "Package", dependency: "Package"
+    ) -> None:
+        """
+        Replace the name and version of a dependency located in a parent directory.
+
+        Go module dependencies referring to a parent directory should be replaced
+        with the parent module name and version. Go package dependencies should
+        have their name replaced with the normalized path with respect to the parent
+        module and version replaced with that of the parent module.
+
+        :param Package package: the package to process
+        :param Package dependency: the dependency to process
+        """
+        if dependency.version and ".." in Path(dependency.version).parts:
+            modules = {
+                package.name: package for package in self.packages if package.type == "gomod"
+            }
+            pkg_module_name = gomod.match_parent_module(package.name, modules.keys())
+            if pkg_module_name is None:
+                # This should be impossible
+                raise RuntimeError(f"Could not find parent Go module for package: {package.name}")
+
+            dep_normpath = os.path.normpath(os.path.join(pkg_module_name, dependency.version))
+            dep_module_name = gomod.match_parent_module(dep_normpath, modules.keys())
+            if dep_module_name is None:
+                raise ContentManifestError(
+                    (f"There is no Go module containing {dep_normpath} in this request")
+                )
+
+            module = modules[dep_module_name]
+            if dependency.type == "gomod":
+                dependency.name = module.name
+                dependency.version = module.version
+            elif dependency.type == "go-package":
+                dependency.name = dep_normpath
+                dependency.version = module.version
```

Comment:
I think this will have to work differently if we want to generate correct purls for `../staging/src/some-module` type of use cases

One idea would be to do roughly this

```python
def process_gomod(self, module, dependency):
    dep_normpath = normpath(join(module.name, dependency.name))
    parent_module = get_parent_module(dep_normpath)
    relpath_from_parent_module_to_dep = relpath(parent_module, dep_normpath)

    dep_purl_with_placeholder = to_purl(dependency, parent_package_relpath=relpath_from...)
    parent_purl = self.get_purl(parent_module)

    dep_purl = replace_parent_purl_placeholder(dep_purl_with_placeholder, parent_purl)
```

But there might be an issue around making sure that you already have/can generate the parent module purl when you need it (or maybe not, haven't thought about it much)

Reply:
Or maybe even simpler than that (I didn't check if `to_purl` does what is needed to achieve this)
```python
dep_purl = f"{parent_purl}#{relpath_from_parent_module_to_dep}"
```